trace-account/src/main.rs

```diff
@@ -0,0 +1,897 @@
+                Ok(data) => data,
+                Err(_) => {
+                    eprintln!("Could not read regids from the provided file.");
+                    return;
+                }
+            };
+            let futures = inputs
+                .iter()
+                .map(|input| trace_single_account(&table, &db, input));
+            future::join_all(futures).await;
+        }
+        Mode::Single {
+            address,
+            decryption_key,
+        } => {
+            let input = match decryption_key {
+                Some(decryption_key) => {
+                    let encryption_secret_key = match hex::decode(&decryption_key)
+                        .context("Hex decoding error")
+                        .and_then(|bs| from_bytes(&mut std::io::Cursor::new(bs)))
+                    {
+                        Ok(v) => Some(v),
+                        Err(e) => {
+                            eprintln!("The provided decryption key is malformed due to: {}", e);
+                            return;
+                        }
+                    };
+                    RetrievalInput {
+                        account_address: address,
+                        encryption_secret_key,
+                    }
+                }
+                None => RetrievalInput {
+                    account_address:       address,
+                    encryption_secret_key: None,
+                },
+            };
+            match trace_single_account(&table, &db, &input).await {
+                Ok(()) => (),
+                Err(e) => panic!("Tracing failed: {}", e),
+            }
+        }
+    };
+}
+
+#[derive(Debug, Serialize)]
+struct CsvRow {
+    time:             String,
+    transaction_type: String,
+    sender:           Option<AccountAddress>,
+    receiver:         Option<AccountAddress>,
+    cost:             Option<Amount>,
+    reward:           Option<Amount>,
+    summed_transfers: Option<AmountDelta>,
+    // summed_transfers + reward - cost
+    public_total:     AmountDelta,
+    // amount added or removed from encrypted funds, when changed
+    encrypted_total:  Option<AmountDelta>,
+}
+
+async fn trace_single_account(
+    table: &elgamal::BabyStepGiantStep<id::constants::ArCurve>,
+    db: &DatabaseClient,
+    input: &RetrievalInput,
+) -> anyhow::Result<()> {
+    let traced_account = input.account_address;
+    let mut file = Path::new(&traced_account.to_string()).to_path_buf(); // todo should we add them to a directory? (Possibly specified by user)
+    file.set_extension("csv");
+    let mut writer =
+        csv::Writer::from_writer(std::fs::File::create(file).expect("Cannot create output file"));
+    let rows = db
+        .query_account(&traced_account, 10000, QueryOrder::Ascending {
+            start: None,
+        })
+        .await?; // todo which limit should be used?
+    println!("Tracing: {}.", traced_account);
+    writer = rows
+        .fold(Ok(writer), |writer, entry| async {
```

Comment:
Splitting into separate tasks where one produces a stream of CsvRows and the other writes them to a file could be faster.